File: src/lib/truncation.ts
/**
 * Convention-based field truncation.
 *
 * Fields named "description", "body", or "content" are truncatable.
 * Truncated fields get a companion `{field}Length` showing full size.
 */

const TRUNCATABLE_FIELDS = new Set(["description", "body", "content"]);
const DEFAULT_MAX_LENGTH = 200;
const ELLIPSIS = "...";

// Module-level state, configured once per CLI invocation
let expandedFields: Set<string> | "all" = new Set();

export function configureTruncation(opts: {
  expand?: string;
  full?: boolean;
}): void {
  if (opts.full) {
    expandedFields = "all";
  } else if (opts.expand) {
    expandedFields = new Set(
      opts.expand.split(",").map((s) => s.trim().toLowerCase()),
    );
  } else {
    expandedFields = new Set();
  }
}

function shouldExpand(fieldName: string): boolean {
  return expandedFields === "all" || expandedFields.has(fieldName);
}

function truncateString(value: string, maxLength: number): string {
  if (value.length <= maxLength) {
    return value;
  }
  return `${value.slice(0, maxLength)}${ELLIPSIS}`;
}

/**
 * Recursively walk data, truncating truncatable fields and adding
 * companion length fields. Works on objects and arrays.
 */
export function applyTruncation(data: unknown): unknown {
  if (data === null || data === undefined) {
    return data;
  }

  if (Array.isArray(data)) {
    return data.map((item) => applyTruncation(item));
  }

  if (typeof data === "object") {
    const obj = data as Record<string, unknown>;
    const result: Record<string, unknown> = {};

    for (const [key, value] of Object.entries(obj)) {
      if (TRUNCATABLE_FIELDS.has(key) && typeof value === "string") {
        result[`${key}Length`] = value.length;
        result[key] = shouldExpand(key)
          ? value
          : truncateString(value, DEFAULT_MAX_LENGTH);
      } else if (typeof value === "object" && value !== null) {
        result[key] = applyTruncation(value);
      } else {
        result[key] = value;
      }
    }

    return result;
  }

  return data;
}

File: internal/truncation/truncation.go
package truncation

import (
	"fmt"
	"strings"
)

const defaultMaxLength = 200
const ellipsis = "\u2026"

var (
	expandedFields map[string]bool
	expandAll      bool
	maxLength      = defaultMaxLength
)

var truncatableFields = map[string]bool{
	"description": true,
	"body":        true,
	"content":     true,
}

// Configure sets truncation state for the CLI invocation.
func Configure(opts ConfigOpts) {
	expandedFields = make(map[string]bool)
	expandAll = false
	maxLength = defaultMaxLength

	if opts.Full {
		expandAll = true
	}
	if opts.Expand != "" {
		for _, f := range strings.Split(opts.Expand, ",") {
			expandedFields[strings.ToLower(strings.TrimSpace(f))] = true
		}
	}
	if opts.MaxLength > 0 {
		maxLength = opts.MaxLength
	}
}

type ConfigOpts struct {
	Expand    string
	Full      bool
	MaxLength int
}

func shouldExpand(field string) bool {
	return expandAll || expandedFields[strings.ToLower(field)]
}

// Apply recursively truncates truncatable fields in the data structure.
// Works on the map[string]any representation produced by JSON marshal/unmarshal.
func Apply(data any) any {
	if data == nil {
		return nil
	}
	switch v := data.(type) {
	case map[string]any:
		out := make(map[string]any, len(v))
		for k, val := range v {
			if truncatableFields[k] {
				if s, ok := val.(string); ok {
					out[fmt.Sprintf("%sLength", k)] = len(s)
					if shouldExpand(k) || len(s) <= maxLength {
						out[k] = s
					} else {
						out[k] = s[:maxLength] + ellipsis
					}
					continue
				}
			}
			out[k] = Apply(val)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = Apply(item)
		}
		return out
	default:
		return data
	}
}
